### PR TITLE
improve REGSAVE_save/restore functions

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

File: src/dmd/backend/cod3.d [hunk 1]
```diff
@@ -208,17 +208,21 @@ private __gshared ubyte[256] inssize2 =
 ];
 
 /*************************************************
- * Allocate register temporaries
+ * Generate code to save `reg` in `regsave`E stack area.
```
Comment (jacob-carlborg) [inline src/dmd/backend/cod3.d]: What’s the trailing `E` in `regsaveE` doing there?

File: src/dmd/backend/cod3.d [hunk 2]
```diff
@@ -230,39 +234,42 @@ void REGSAVE_save(ref REGSAVE regsave, ref CodeBuilder cdb, reg_t reg, uint *pid
...
     reflocal = true;
     if (regsave.idx > regsave.top)
         regsave.top = regsave.idx;              // keep high water mark
-    *pidx = i;
 }
 
-void REGSAVE_restore(ref REGSAVE regsave, ref CodeBuilder cdb, reg_t reg, uint index)
+/*******************************
+ * Restore `reg` from `regsave` area.
+ */
```
Comment (jacob-carlborg) [inline src/dmd/backend/cod3.d]: Would be nice to flesh out this documentation with a params section as well.

File: src/dmd/backend/cod3.d [hunk 3]
```diff
@@ -208,17 +208,21 @@ private __gshared ubyte[256] inssize2 =
...
+ *      reg = register to save
+ *      idx = set to location in regsave for use in REGSAVE_restore()
  */
 
-void REGSAVE_save(ref REGSAVE regsave, ref CodeBuilder cdb, reg_t reg, uint *pidx)
+void REGSAVE_save(ref REGSAVE regsave, ref CodeBuilder cdb, reg_t reg, out uint idx)
 {
-    uint i;
-    if (reg >= XMM0)
+    if (isXMMreg(reg))
```
Comment (jacob-carlborg) [inline src/dmd/backend/cod3.d]: This will have slightly different behavior but it might be more correct.

Comment (WalterBright) [reply src/dmd/backend/cod3.d]: Not only is it more correct, it expresses the intent.

**Review by RazvanN7 — APPROVED** [review]

File: src/dmd/backend/cod3.d (hunk 2 above)
Comment (RazvanN7) [reply src/dmd/backend/cod3.d]: It says that it is complement to `REGSAVE_save` so it can be deduced that is has the same parameters

**Pull request merged**